test/GPv2Settlement.test.ts

```diff
@@ -398,6 +421,89 @@ describe("GPv2Settlement", () => {
       });
     });
 
+    describe("Order Executed Fees", () => {
+      const { sellAmount, buyAmount } = partialOrder;
+      const feeAmount = ethers.utils.parseEther("10");
+      const { [sellToken]: sellPrice, [buyToken]: buyPrice } = prices;
+      const computeInTransferForOrderVariant = async (
+        { kind, partiallyFillable }: Pick<Order, "kind" | "partiallyFillable">,
+        executedAmount?: BigNumber,
+      ) => {
+        const encoder = new SettlementEncoder(testDomain);
+        await encoder.signEncodeTrade(
+          {
+            ...partialOrder,
+            feeAmount,
+            kind,
+            partiallyFillable,
+          },
+          executedAmount || 0,
```

Comment:
This looked weird at first, because I was expecting that if the executed amount wasn't specified then you'd execute the full amount. This was incorrect, because the idea is that if you don't bother to specify the executed amount then you want it to be zero to save gas in fill or kill orders.
So I wondered: would it be useful/possible to do some Typescript magic that enforces that `executedAmount` must be defined if and only if `partiallyFillable == true`? This would be nice also in `signEncodeTrade`. (Maybe not worth it if we plan to use RLP order encoding in the future.)

Reply:
> So I wondered: would it be useful/possible to do some Typescript magic that enforces that executedAmount must be defined if and only if partiallyFillable == true

There is... the reason I didn't use it is the error message around it is very bad:
```ts
function foo({ kind, partiallyFillable }: { kind: string, partiallyFillable: false }): void;
function foo({ kind, partiallyFillable }: { kind: string, partiallyFillable: true }, executedAmount: number):void ;
function foo({ kind, partiallyFillable }: { kind: string, partiallyFillable: boolean }, executedAmount?: number): void {
    console.log(kind, partiallyFillable, executedAmount);
}

foo({ kind: "sell", partiallyFillable: false }); // OK
foo({ kind: "sell", partiallyFillable: true }, 10); // OK
foo({ kind: "sell", partiallyFillable: true }); // ERR: Type 'true' is not assignable to type 'false'.
```

Try it out [here](https://www.typescriptlang.org/play?ts=4.1.2#code/GYVwdgxgLglg9mABMOcAUBvRBrGYAmANIgA4CGATrGQDY0CeAYjHWQEY0CmiAvgFyIsuAgIDOUCngDmxclRi0GzVh04DgtUdx4BKAQDc4MfAG4AUKEiwEyVJhx4ipStTpMWNdl14ChjsRLSsi4KbsqeqgISINrEnAAenBAgUJz4AIIAtnDgUAJgIJlsnBR6hsaI5pbQ8Ego6H4EwfKK7ire-IIOIojikmAyzi1hHl5qiGyoXGRIPHGJyakZ2bkA-PmFxaUGRviCZoiHiBAIonBcAHQ0cFJowk5yrkqjqvNJKWlZOWBQOuY8Zgsdka+AEACItHQwc0nm0Ilx1JptH9EAB6VGIADyAGkgQ1uqDEBDOFCYaFnu1xtFYogAIwABhR6KxuPq9nu4MhNGhQ1h4TGUQoMV4TIxAFEAEoSgQAFXoJG4AHJqYrEDBRIgwHAoIgyKJRDApGAxogoHBTfKlRoaFpFRcgA)